agent/api/retry_handler.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2014-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package api
+
+import (
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/utils"
+	"github.com/aws/amazon-ecs-agent/agent/utils/ttime"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+)
+
+const maxSubmitRetryDelay = 5 * time.Minute
+const submitRetryDelayJitter = 3 * time.Minute
+
+// maxSubmitRetryCount is how many times it will try to send the
+// Submit*StateChange results on retriable errors before giving up forever.
+// The below number was arrived at by estimating the number for retrying for 24 hours.
+// The base retry handler does 2*n * 50 milliseconds of sleep.
+// First we figure out how many retries it needs to reach 5 minutes:
+// 5 min = \sum_{i=0}^n{2^i * 50 ms}
+// 300000 ms = \sum_{i=0}^n{2^i} * 50ms
+// 6000 = 2^{n+1} - 1
+// n ~= 11.5
+// So the first 12 tries will take just over 5 minutes, meaning that to get 24 hours we have:
+// 24 hours ~= 12 baseTries * 5min / 12 baseTries + n extraTries * 5min/extraTry
+// extraTries = 288
+// retries = 12 + 288
+const maxSubmitRetryCount = 12 + 288
+
+const opSubmitContainerStateChange = "SubmitContainerStateChange"
+const opSubmitTaskStateChange = "SubmitTaskStateChange"
+
+// ECSRetryHandler defines how to retry ECS service calls. It behaves like the default retry handler, except for the SubmitStateChange operations where it has a massive upper limit on retry counts
+func ECSRetryHandler(r *aws.Request) {
+	if r.Operation != nil && (r.Operation.Name == opSubmitContainerStateChange || r.Operation.Name == opSubmitTaskStateChange) {
```

Comment:
For readability, I'd rather that you reversed this conditional:
```
if r.Operation == nil || (r.Operation.Name != opSubmitContainerStateChange && r.Operation.Name != opSubmitTaskStateChange) {
	aws.AfterRetryHandler(r)
	return
}
// the rest of the special-cased retry code here
```
That moves you to exit early when you don't need anything special and reduces your indentation by 1 for the special-cased code.